src/Symfony/Component/DependencyInjection/Compiler/ResolveInstanceofConditionalsPass.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\ChildDefinition;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+
+/**
+ * Applies instanceof conditionals to definitions.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class ResolveInstanceofConditionalsPass implements CompilerPassInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        $didProcess = false;
+        foreach ($container->getDefinitions() as $id => $definition) {
+            if ($definition instanceof ChildDefinition) {
+                continue;
+            }
+            if ($definition !== $processedDefinition = $this->processDefinition($container, $id, $definition)) {
+                $didProcess = true;
+                $container->setDefinition($id, $processedDefinition);
+            }
+        }
+        if ($didProcess) {
+            $container->register('abstract.'.__CLASS__, '')->setAbstract(true);
```

Comment:
that's the base abstract service all "instanceof" chains inherit from. See [L56](https://github.com/symfony/symfony/pull/22356/files#diff-af848a09080e51b3fcccb86f20d716a9R56) below.